Add new subcommand `npm doctor`

watilde opened the pull request:
From here https://github.com/npm/npm/issues/6756#issuecomment-251798653

*this screenshot was updated with current progress at 4th December 2016*
![screenshot](https://gyazo.com/d129f84696c972fdaf580a5b0deca4b5.png)

ToDo:
- [x] Add new subcommand
- [x] Write a test
- [x] Replace `getLatestNodejsVersion` with a fixturable way
- [x] Update docs

watilde committed:
util: put ansiTrim into util from outdated

legodude17 commented:
Does this add any new deps?

othiym23 commented:
@watilde I owe you a response! I'll try to get to it today! The summary is this looks fantastic, but we need to talk a lil more about the functionality everybody wants.

legodude17 commented:
Current comments:
+ Any ideas on a more exhaustive test then, "It prints things out with 5 lines"? I was thinking maybe make sure it prints a table and that the columns match.
+ What is the best way to get the latest `node` version? I think that yours is the best way.

watilde commented:
@legodude17 Yes, that's a good point. Before that, I'm thinking about a way to make the test having a fixture in test.

@othiym23 Sure thing! Let's continue the discussion at https://github.com/npm/npm/issues/6756

legodude17 commented:
@watilde is this still `in-progress`?

watilde commented:
@legodude17 It's fixed technically for the original plan, but the code potentially will be changed depending on what kind of request comes from the issue thread. Then the status doesn't matter anymore :)

legodude17 commented:
One suggestion from #6756 is to offer to fix it like `Would you like me to run npm cache clean for you?`. I think that would be great, but maybe in a separate PR.

othiym23 reviewed (changes requested):
@watilde, my apologies for the lengthy delay. It's been a bruiser of a month.

Here, in no particular order, are my thoughts:

1. I'm not in love with the idea of prescribing that users nuke their caches every time they run `npm doctor`. It makes installs (potentially much) slower until the cache rewarms, and in cases in which the user is on a flaky internet connection, can actually make installs even flakier. If at all possible, I'd like to move to advising people to only `npm cache clean` individual broken dependencies.

    It's probably out of scope for this initial spike, but something that validates that the cache is in good shape (and maybe went ahead and cleaned up any shasum mismatches automatically) would be a super useful alternative.
2. Something that validates that `$HOME/.npm` and all its children is owned and writable by `$USER` would be very helpful for diagnostic purposes.
3. Similarly, something that checks whether `join(npm.config.globalPrefix, 'lib', 'node_modules')` is writable and tells the user whether they need to use `sudo` to run global installs would be very helpful.
4. It would be _very_ helpful to verify that Git is installed and available, especially for Windows users.
5. This is another thing that doesn't have to be added yet, but something that builds a very simple canned native module with `node-gyp` would be extremely helpful for Windows users unsure whether they have their toolchain correctly installed.
6. Finally, and longest-term, it would be _fantastic_ to have some way for npm to verify its own integrity, which probably requires that some kind of install manifest be generated for it. While it would be useful to have this functionality for npm packages in general, it's especially important for npm, because of the Node.js installer's bad habit of just copying the new npm right over the top of the old one, which causes a bunch of the stranger `module not found` bugs that show up on the issue tracker. Being able to catch this automatically would be hugely useful.

This is basically my wishlist, though. What you've got is a fantastic start, and when you get the above comments addressed, I'll ship what you've got so far. Thank you so much for putting this together!

legodude17 commented on lib/doctor.js:
```diff
@@ -0,0 +1,112 @@
+
+  chain(actionsToRun, function (stderr, stdout) {
+    if (stderr) return cb(stderr)
+    var outHead = ['Question', 'Answer', 'Prescription']
```
Good point @othiym23.

legodude17 commented on doc/cli/npm-doctor.md:
```diff
@@ -0,0 +1,17 @@
+
+This command helps you when you get struck by environments which
+potentially have a problem. Some of the questions are same with
+we ask in the ISSUE_TEMPLATE.md.
```
I think the best way to do that would be to template it, but I am not sure about what anyone's thoughts are about including a template engine with `npm`.

watilde replied:
That's absolutly good idea, and I'd like to have the feature in `npm report` command which have been proposaling at https://github.com/npm/npm/issues/5214. I'm planning to use the `ISSUE_TEMPLATE.md` as a acutual template, but let's separete the reporting part from this diagnosis part to make the initial scope clear and simple.

legodude17 commented on lib/doctor.js:
```diff
@@ -0,0 +1,112 @@
+
+function getLatestNpmVersion (cb) {
+  fetchPackageMetadata('npm', '.', function (er, d) {
+    cb(er, d.version)
```
So `cb(er, d['dist-tags'].latest)`?

watilde commented on test/tap/doctor.js:
```diff
@@ -0,0 +1,29 @@
+})
+
+test('npm doctor', function (t) {
+  npm.commands.doctor({url: 'http://localhost:1337/'}, true, function (err, list) {
```
That's right! I've updated.

watilde commented:
@othiym23 Hey Forrest, thanks for your thoughtful wish list! I'd like to separate examination processes and auto-fix features and let me reply to your every wish-item.

1. Agreed! This request is the rest of the things I'd like to put into this PR.
2. I've added into this PR already.
3. I've added into this PR already.
4. I've added into this PR already.
5. I'd like to have [node-gyp doctor](https://github.com/nodejs/node-gyp/issues/1048) instead of putting into `npm doctor`. Let's take out this from the current initial scope.
6. This request may make this PR turning into a giant case of scope creep like you said. Could we also take out this from the current initial scope?

Additionally, I've added more permission check for `join(npm.config.localPrefix, 'lib', 'node_modules')` since I was missing! As the initial plan to iterate from it, let's start with the following check-list as the smallest subset of information necessary to improve the supportability of npm.
1. npm ping
2. npm -v
3. node -v
4. npm config get registry
5. which git
6. Perms check on cached files
7. Perms check on `join(npm.config.globalPrefix, 'lib', 'node_modules')`
8. Perms check on `join(npm.config.localPrefix, 'lib', 'node_modules')`

Let me ping you when they're ready.

legodude17 commented:
Some random ideas for things to add:
  1. Ensure the proxy is setup correctly (somehow)
  2. Ensure git is present and works

Keep up the great work @watilde!

watilde commented on test/tap/doctor.js:
```diff
@@ -0,0 +1,29 @@
+test('npm doctor', function (t) {
+  npm.commands.doctor({url: 'http://localhost:1337/'}, true, function (err, list) {
+    t.ifError(err, 'npm loaded successfully')
+    t.same(list.length, 5, 'list should have 5 prop')
```
Yeah it should be since `npm doctor` ask the things to the registry through `npm ping` and `fetchPackageMetadata`. I've added `npm-registry-mock` into this test and verifying all of the output!

watilde commented:
@othiym23 I've added some check items that users need, and written up the detailed documentation for the `npm doctor` command! Could I possibly request your review again?

(๑˃̵ᴗ˂̵)و let me cc to you as well @zkat

watilde committed:
doctor: add new subcommand

This command will diagnose user's environment and let
the user know some recommended solutions if they
potentially have any problems related to npm.

othiym23 commented:
I think this will be ready in time to include in this week's release, which might end up being the last release of the CLI in 2016! I only have a few further comments:

1. Right now, the command seems to hang while it's running (I'm going to guess most of that is due to checking the shasums in the cache). If you add a log statement to each check at loglevel info and enable the progress bar, this will let users know that in fact something is happening.
2. Because validating the hashes can be very time-consuming if people have a lot of packages in their cache, it might make sense to log at level `debug` each shasum you're validating, especially because you're doing so synchronously. Because the completion routine returns the number of files to check, it should be pretty easy to set up a progress group that will track the process of validating the cache. (This is a nice to have, and if it turns out to be tricky, don't worry about it and somebody can add it in later.)
3. In the next day or two, I'll make a pass at expanding the docs you've written thus far, and make sure they'll tie in nicely with the the new `TROUBLESHOOTING.md` that @snopeks and @zkat are in the process of assembling.

This is really great! I can't wait to get it out into people's hands. Thank you so much for your patience and careful work, Daijiro!

iarna commented on lib/doctor.js:
```diff
@@ -0,0 +1,107 @@
+  var nodeLTS = p[2].replace('v', '')
+  var whichGit = p[3] || 'not installed'
+  var readbleCaches = p[4] ? 'ok' : 'notOk'
+  var executableGlobalModules = p[4] ? 'ok' : 'notOk'
```
Shouldn't these from this line down be 5, 6 & 7?

iarna commented on lib/doctor/checksum-cached-files.js:
```diff
@@ -0,0 +1,35 @@
+      if (!/.tgz$/.test(file)) return
+      var tgz = fs.readFileSync(file)
+      try {
+        var pkgJSON = fs.readFileSync(path.join(file, '../package/package.json'))
```
I think what's wanted here is:

`path.join(path.dirname(file), 'package/package.json')`

iarna commented on lib/doctor/checksum-cached-files.js:
```diff
@@ -0,0 +1,35 @@
+      var tgz = fs.readFileSync(file)
+      try {
+        var pkgJSON = fs.readFileSync(path.join(file, '../package/package.json'))
+        var shasum = JSON.parse(pkgJSON).dist.shasum
```
`dist` is added by the registry and won't be present for modules installed from other sources. `_shasum` should be available however.

iarna commented on lib/doctor/check-files-permission.js:
```diff
@@ -0,0 +1,38 @@
+
+function checkFilesPermission (root, mask, cb) {
+  var accessible = true
+  if (process.platform === 'win32') return cb(null, !accessible)
```
This should probably be calling back with `true` shouldn't it? We don't want it to mark windows as always failing right?

iarna commented on lib/doctor.js:
```diff
@@ -0,0 +1,107 @@
+    [getLatestNpmVersion],
+    [getLatestNodejsVersion, args['node-url']],
+    [which, 'git'],
+    [checkFilesPermission, npm.cache, 4],
```
`filemode & 0400` would be true for files you have read access to. Should this be rewrite instead? (`6`)

iarna commented on lib/doctor.js:
```diff
@@ -0,0 +1,107 @@
+    [getLatestNodejsVersion, args['node-url']],
+    [which, 'git'],
+    [checkFilesPermission, npm.cache, 4],
+    [checkFilesPermission, globalNodeModules, 1],
```
`filemode & 0100` is true if you have execute access to. I'm assuming that this is supposed to be checking for read access? (`4`)

iarna committed:
fixup: add progress bars, fix review bugs (to be squashed)

iarna commented:
I put together a little patch that fixes the bugs that I encountered and reworks things to allow progress bars. You can see that here:

https://github.com/npm/npm/commit/7cc791df91bc66446f410711c9058763df41ca26

watilde commented:
@iarna Thanks for your comments, and your patch seems so brilliant.. Let me do cherry-pick it into this PR, or you also can handle the commits to be suited for next-release.

iarna commented:
Thanks again for this! It was merged and released in 4.1.0.
